kafka-client/src/main/java/dev/responsive/kafka/internal/db/partitioning/SegmentPartitioner.java

```diff
@@ -0,0 +1,194 @@
+ * <p>
+ * Each segment covers a specific time range in milliseconds. If t_n is the start time of the
+ * nth segment, then that segment covers all timestamps from t_s - t_s + segmentInterval.
+ * For each segment id there are N physical table partitions, where N is the number of kafka
+ * partitions. Each physical state store instance should only be interacting with a single kafka
+ * partition, but will generally have multiple active segments for each kafka partition.
+ * <p>
+ * For a given kafka partition, the total number of segments/table partitions is determined by
+ * the store-level config for num_segments, which controls the width of each segment according
+ * to the relationship num_segments * segmentInterval = retentionPeriod.
+ * Note: the actual number of physical segments at any given time is likely to be num_segments + 1,
+ * for two reasons. First, if the configured numSegments does not evenly divide the retentionPeriod,
+ * an additional partial segment will always be required to cover any remainder of the full
+ * retention period left. Second, at any given point in time it's likely that the time range
+ * covered by the oldest segment is partially expired, while the upper end of the newest segment
+ * may not yet be active/filled.
+ * <p>
+ * When stream-time for a given kafka partition advances, a tombstone is sent to delete the entire
+ * oldest physical segment (or segments, if it has advanced by more than the segmentInterval).
+ * At the same time, a new physical segment (or segments) will be created and the corresponding
+ * remote partition initialized.
+ * <p>
+ * Let's look at an example case, at a particular moment in time, with the following configuration:
+ * N = 4
+ * retentionPeriod = 100ms
+ * numSegments = 3
+ * segmentInterval = 33
+ * <p>
+ * kafkaPartition | stream-time | minValidTs |    active segment time bounds   |  segmentId ||
+ *           0    |     16      |      0     | 0-32                            | 0          ||
+ *           1    |     88      |      0     | 0-32, 33-65, 66-98              | 0, 1, 2    ||
+ *           2    |     101     |      2     | 0-32, 33-65, 66-98, 99-131      | 0, 1, 2, 3 ||
+ *           3    |     169     |      70    | 66-98, 99-131, 132-164, 165-197 | 1, 2, 3, 4 ||
+ * <p>
+ * NOTE: because we are already dividing up the partition space into segments, we don't further
+ * split things into true sub-partitions based on key. Each kafka partition still maps to multiple
+ * table partitions which should help with the parallelism, but unlike the subpartitioner, the
+ * segment-partitioning scheme is not static and temporal rather than key-based. If data is skewed
+ * in time, this might result in uneven partitions and a need to further subdivide the partition
+ * space.
+ * <p>
+ * For the time being, we simply recommend that users configure the number of segments
+ * similarly to how they would configure the number of sub-partitions for a key-value store.
```

Comment:
is this feasible? experimentally I needed the number of sub-partitions to be in the thousands for decent performance, but then you may end up scanning many segments to get the range you need (may be OK)

Reply:
Guess it depends on what kind of range scans people are actually doing?

Keep in mind, even though I did implement the basic single-key range scan in this PR, only the point lookup is actually used by any DSL operators. The range scan is used for stream-stream joins but we really just need a completely different table implementation for those (maybe not *completely* different, but certainly separate). I've thought a little bit about the stream-stream join implementation and I think in that case we could just do a little math based on the window size, since that determines the length of the range scan.

But I'm getting ahead of myself. If we just focus on the point lookups, we probably want a similarly large number of partitions, whether they are segments or subpartitions, right?

I guess some experimentation is needed here, and we can always recommend users to select a smaller number of segments if they plan to use range scans? This is an internal class so it's unlikely users will see these comments anyways, happy to remove the comment for now or just leave it up until we have something more concrete to say. Either way we definitely need to write a docs page for the window store once it's fully production ready